scanner/services/matcher.go

```diff
@@ -16,18 +23,83 @@ import (
 // matcherService represents a vulnerability matcher gRPC service.
 type matcherService struct {
 	v4.UnimplementedMatcherServer
+	// indexer is used to retrieve index reports.
+	indexer indexer.Indexer
+	// matcher is used to match vulnerabilities with index contents.
 	matcher matcher.Matcher
+	// disableEmptyContents allows the vulnerability matching API to reject requests with empty contents.
+	disableEmptyContents bool
 }
 
-// NewMatcherService creates a new vulnerability matcher gRPC service.
-func NewMatcherService(matcher matcher.Matcher) *matcherService {
+// NewMatcherService creates a new vulnerability matcher gRPC service, to enable
+// empty content in enrich requests, pass a nil indexer.
+func NewMatcherService(matcher matcher.Matcher, indexer indexer.Indexer) *matcherService {
 	return &matcherService{
-		matcher: matcher,
+		matcher:              matcher,
+		indexer:              indexer,
+		disableEmptyContents: indexer == nil,
 	}
 }
 
-func (s *matcherService) GetVulnerabilities(_ context.Context, _ *v4.GetVulnerabilitiesRequest) (*v4.VulnerabilityReport, error) {
-	return nil, status.Error(codes.Unimplemented, "method GetVulnerabilities not implemented")
+func (s *matcherService) GetVulnerabilities(ctx context.Context, req *v4.GetVulnerabilitiesRequest) (*v4.VulnerabilityReport, error) {
+	ctx = zlog.ContextWithValues(ctx, "component", "scanner/service/matcher")
+	if err := validators.ValidateGetVulnerabilitiesRequest(req); err != nil {
+		return nil, errox.InvalidArgs.CausedBy(err)
+	}
+	// Get an index report to enrich: either using the indexer, or provided in the request.
+	var ir *claircore.IndexReport
+	var err error
+	if req.GetContents() == nil {
+		if s.disableEmptyContents {
+			zlog.Debug(ctx).Msg("no contents, rejecting")
```

Comment:
Will the logs contain any identifier that will help match to the request?  (ie: should we log the `hashId` or similar?)

If the logs resemble what is below then may not be helpful outside of a local / testing use case.
```json
{"component": "scanner/service/matcher", "msg": "no contents, rejecting"}
{"component": "scanner/service/matcher", "msg": "no contents, rejecting"}
{"component": "scanner/service/matcher", "msg": "no contents, retrieving"}
{"component": "scanner/service/matcher", "msg": "no contents, parsing"}
```

Reply:
> Will the logs contain any identifier that will help match to the request?

I considered including more details, but I believe a simpler solution would be to generate a unique identifier for each request, include it in the context, and log the input as JSON once the request is admitted. This will allow us to easily track that specific input in the logs without repeating the same information. 

Do you agree? 

I have a task that would be appropriate for this approach.

Reply:
Yea am ok with using a request ID / tracing.

Reply:
> I have a task that would be appropriate for this approach.

Is this a Jira ticket? If so, which one?